torch/csrc/autograd/record_function.cpp

```diff
@@ -145,12 +148,28 @@ void runBeforeCallbacks(RecordFunction* rf, const std::string& funcName) {
   if (hasCallbacks()) {
     auto run_samples = shouldRunSampledCallbacks();
     if (run_samples || hasNonSampledCallbacks()) {
-      rf->setRunSampled(run_samples);
+      rf->_setRunSampled(run_samples);
       rf->before(funcName);
     }
   }
 }
 
+void RecordFunction::_setCurrent() {
+  parent_ = thread_local_func_;
+  thread_local_func_ = this;
+  is_current_ = true;
+}
+
+/* static */
+uint16_t RecordFunction::getCurrentThreadId() {
+  if (!current_thread_id_) {
+    // happens only once per thread
+    std::lock_guard<std::mutex> guard(next_thread_id_mutex_);
```

Comment:
Are we sure that we can't deadlock in any case by different threads acquiring `getCurrentThreadId` and requiring the `all_event_list_map_mutex` in different orders?

Reply:
we don't acquire locks under next_thread_id_mutex_ so deadlock is not possible